src/protocol/taiko_alethia/ProverManager.sol

```diff
@@ -0,0 +1,405 @@
+        inbox = _inbox;
+        checkpointTracker = ICheckpointTracker(_checkpointTracker);
+        publicationFeed = IPublicationFeed(_publicationFeed);
+
+        // Initialize the first period with a known prover and a set fee
+        require(msg.value >= _config.livenessBond, "Insufficient balance for liveness bond");
+        periods[0].prover = _initialProover;
+        periods[0].stake = _config.livenessBond;
+        periods[0].fee = _initialFee;
+    }
+
+    /// @notice Deposit ETH into the contract. The deposit can be used both for opting in as a prover or proposer
+    function deposit() external payable {
+        balances[msg.sender] += msg.value;
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    /// @notice Withdraw available(unlocked) funds.
+    /// @param amount The amount to withdraw
+    function withdraw(uint256 amount) external {
+        balances[msg.sender] -= amount;
+
+        (bool ok,) = payable(msg.sender).call{value: amount}("");
+        require(ok, "Withdrawal failed");
+        emit Withdrawal(msg.sender, amount);
+    }
+
+    /// @inheritdoc IProposerFees
+    /// @dev This function advances to the next period if the current period has ended.
+    // TODO: deal with delayed publications
+    function payPublicationFee(address proposer, bool isDelayed) external payable {
+        require(msg.sender == inbox, "Only the Inbox contract can call this function");
+
+        // Accept additional deposit if sent
+        if (msg.value > 0) {
+            balances[proposer] += msg.value;
+            emit Deposit(proposer, msg.value);
+        }
+
+        uint256 _currentPeriod = currentPeriodId;
+
+        uint256 periodEnd = periods[_currentPeriod].end;
+        if (periodEnd != 0 && block.timestamp > periodEnd) {
+            // Advance to the next period
+            currentPeriodId++;
+            _currentPeriod++;
+            emit NewPeriod(_currentPeriod);
+        }
+
+        uint256 requiredFee = periods[_currentPeriod].fee;
+
+        // Deduct fee from proposer's balance and add to accumulated fees
+        balances[proposer] -= requiredFee;
+        periods[_currentPeriod].accumulatedFees += requiredFee;
+    }
+
+    /// @inheritdoc IProverManager
+    /// @dev The offered fee has to be at least `minUndercutPercentage` lower than the current best price.
+    /// @dev The current best price may be the current prover's fee or the fee of the next bid, depending on a few
+    /// conditions.
+    function bid(uint256 offeredFee) external {
+        uint256 _livenessBond = livenessBond;
+        require(balances[msg.sender] >= _livenessBond, "Insufficient balance for liveness bond");
+
+        uint256 currentPeriod = currentPeriodId;
+        Period storage _currentPeriod = periods[currentPeriod];
+        Period storage _nextPeriod = periods[currentPeriod + 1];
+        uint256 requiredMaxFee;
+        if (_currentPeriod.end == 0) {
```

Comment:
I think we should wrap `_currentPeriod.end == 0` in an `_isPeriodActive(uint256 end)`  internal function. If so:

```suggestion
        if (_isPeriodActive(_currentPeriod.end)) {
```

The core idea is that the contract API should allow developers customizing the rollup to bring their own definitions of what each piece of the contract means.

It's also true that this contract is specific to Taiko. Maybe is it worth discussing the level of customizability we want to allow and where?